prusti-encoder/src/encoders/type/predicate.rs

```diff
@@ -186,48 +173,111 @@ impl<'vir> PredicateEncOutputRef<'vir> {
         vid.map(|vid| self.expect_variant(vid).predicate)
             .unwrap_or(self.ref_to_pred)
     }
-    #[track_caller]
-    pub fn expect_variant_opt(
+
+    pub fn get_ref_to_pred(
         &self,
         vid: Option<abi::VariantIdx>,
-    ) -> &PredicateEncDataStruct<'vir> {
+    ) -> PredicateIdn<'vir, (vir::Ref, vir::ManyTyVal)> {
+        vid.map(|vid| {
+                let data = self.expect_enumlike().expect("empty enum");
+                data.variants[vid.as_usize()].predicate
+        }).unwrap_or(self.ref_to_pred)
+    }
+
+    /// Returns `None` if the `vid` is `None` and this is an enum.
+    pub fn get_variant_opt(
+        &self,
+        vid: Option<abi::VariantIdx>,
+    ) -> Option<&PredicateEncDataStruct<'vir>> {
         match vid {
-            None => self.expect_structlike(),
+            None => self.get_structlike(),
             Some(vid) => {
-                &self.expect_enumlike().expect("empty enum").variants[vid.as_usize()].fields
+                Some(&self.expect_enumlike().expect("empty enum").variants[vid.as_usize()].fields)
             }
         }
     }
 }
 
 pub(crate) struct PredicateBuilder<'vir> {
+    pub(crate) generic_decls: Vec<vir::LocalDeclTyVal<'vir>>,
+    pub(crate) generic_exprs: Vec<vir::ExprTyVal<'vir>>,
+    pub(crate) generic_tys: &'vir [vir::TypeTyVal<'vir>],
+
+    pub(crate) inner: PredicateBuilt<'vir>,
+}
+
+pub(crate) struct PredicateBuilt<'vir> {
```

Comment:
Rename? Especially since this is the representation the builder uses as the build is in progress.